contracts/chain-adapters/l2/Ovm_WithdrawalAdapter.sol

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+pragma solidity ^0.8.0;
+
+import { WithdrawalAdapterBase } from "./WithdrawalAdapterBase.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { WETH9Interface } from "../../external/interfaces/WETH9Interface.sol";
+import { ITokenMessenger } from "../../external/interfaces/CCTPInterfaces.sol";
+import { Lib_PredeployAddresses } from "@eth-optimism/contracts/libraries/constants/Lib_PredeployAddresses.sol";
+import { IL2ERC20Bridge } from "../../Ovm_SpokePool.sol";
+
+/**
+ * @notice Minimal interface for the Ovm_SpokePool contract. This interface is called to pull state from the network's
+ * spoke pool contract to be used by this withdrawal adapter.
+ */
+interface IOvm_SpokePool {
+    // Returns the address of the token bridge for the input l2 token.
+    function tokenBridges(address token) external view returns (address);
+
+    // Returns the address of the l1 token set in the spoke pool for the input l2 token.
+    function remoteL1Tokens(address token) external view returns (address);
+
+    // Returns the address for the representation of ETH on the l2.
+    function l2Eth() external view returns (address);
+
+    // Returns the address of the wrapped native token for the L2.
+    function wrappedNativeToken() external view returns (WETH9Interface);
+
+    // Returns the amount of gas the contract allocates for a token withdrawal.
+    function l1Gas() external view returns (uint32);
+}
+
+/**
+ * @title Ovm_WithdrawalAdapter
+ * @notice This contract interfaces with L2-L1 token bridges and withdraws tokens to a single address on L1.
+ * @dev This contract should be deployed on OpStack L2s which both have a Ovm_SpokePool contract deployed to the L2
+ * network AND only use token bridges defined in the Ovm_SpokePool. A notable exception to this requirement is Optimism,
+ * which has a special SNX bridge (and thus this adapter will NOT work for Optimism).
+ * @custom:security-contact bugs@across.to
+ */
+contract Ovm_WithdrawalAdapter is WithdrawalAdapterBase {
```

Comment:
Should this be upgradeable? If not then is the process for upgrading this and then updating the withdrawTo address in the L3 spoke pool easy?

Reply:
After discussing w/ @bmzig:
- We're probably OK w/o upgradability here. To change this implementation we'd need to deploy a new instance and call `setHubPool()` on the L3 Spoke, which should be fine.
- We are probably lacking a way to rescue any tokens from this contract in case of an implementation problem in `withdrawToken()`. We probably want to support an L1 -> L2 message initiated by the HubPool that instructs this contract to withdraw tokens to an L2 address. We could use this to withdraw tokens to the relevant SpokePool on the same chain and evacuate the tokens via there.

Reply:
This, along with a basic unit test, was added here https://github.com/across-protocol/contracts/pull/605/commits/da1bc9abb9f9f75e83652b7e9b5be7ba41e1dca3